cache_cleaning_scripts: Expand cache cleaning scripts so that they can be used in prod as well.

HarshitOnGitHub opened the pull request:
This PR is an effort to expand all the three cache cleaning scripts by making them aware of the deployments directory. This includes:
* Complete re-write of all the three cache cleaning scripts.
* Moving them from `tools/` to `scripts/` folder so that they are accessible in prod.
* Addition of a new `clean-unused-caches` script which can be used to run all the three scripts in one go(useful mainly for dev env).
* Complete rewrite of `purge-old-deployments` script to make it more configurable in terms of the threshold days and to also make it clean all the orphaned caches automatically.

Fixes: #5726.

timabbott commented:
The NPM one is buggy, in that it seems to want to clean the directory under `/srv/zulip/npm-cache` where we put `yarn` itself:
> Cleaning unused yarn cache /srv/zulip-npm-cache/yarn

timabbott commented:
Thanks for working on this @HarshitOnGitHub!  I merged through "clean-venv-cache: Expand `clean-venv-cache` tool.", with a few changes:
* Line-wrapped the changes in `zulip_tools.py` better
* Remove the unnecessary first print() statement.

The `node_cache.py` change feels a bit wrong; you at the very least should update the type annotation to handle returning `None`, but I think actually better would be to have the check for whether it's a valid path happen elsewhere than `generate_sha1sum_node_modules`, and have that function just be responsible for hashing.  Seems like it'd have cleaner error handling.

A few larger changes you should make throughout:
* The `main()` methods of the various scripts are highly duplicated.  I think we could move that code into a function that just takes a few arguments: the directory, the `get_caches_in_use` function, and the name of the cache to use in print statements.
* In several places, you're doing `subprocess.check_call` when it could instead import the other file and just call its functions.  E.g. `hash_reqs.py`.  I think you could make the `clean-venv-cache` script significantly faster if you did the import approach; can you investigate that?

HarshitOnGitHub commented:
>The NPM one is buggy, in that it seems to want to clean the directory under /srv/zulip/npm-cache where we put yarn itself:
>>Cleaning unused yarn cache /srv/zulip-npm-cache/yarn

@timabbott I think we put yarn in `/srv/zulip-yarn`?

HarshitOnGitHub commented:
>The main() methods of the various scripts are highly duplicated. I think we could move that code into a function that just takes a few arguments: the directory, the get_caches_in_use function, and the name of the cache to use in print statements.

I was also thinking but was not sure if having a generic with that many arguments would be readable and that is why I dropped the idea.

HarshitOnGitHub commented:
One more thing that I forgot to ask you is that the approach we used for determining which venv caches are in use in the previous script seems to be a bit buggy to me in that if a dependency gets upgrade without bumping the provision version(the user isn't forced to reprovision), hash of the venv being used currently and one obtained by hashing the requirements file is different which marks the current cache being used for cleanup which breaks the dev environment. I think we should not delete the cache currently being used. I had thought of asking you about this since you had made this change in commit 404c61ba97bfa322be0bb7ffaf610411a519547c but forgot so bringing up this now.(In the venv commit merged I have used the same approach)

timabbott commented:
Ahh, yes, on the development side, we should include the currently active one as well, for precisely the reason you describe.

timabbott commented:
Awesome, thanks for all your work on this @HarshitOnGitHub!  I just merged everything but the last commit, with a few small tweaks.

For the last commit, I'm not sure, but it seems like we might be able to make `purge-old-deployments` use some of the same cache-cleaning library code with a result of it being more readable; did you consider doing that?

timabbott commented:
I also merged 1a1df290538da74e5664fce0b633bcc375906063, which fixes a small issue I found testing this out on chat.zulip.org.

HarshitOnGitHub commented:
@timabbott I have updated this PR!


>For the last commit, I'm not sure, but it seems like we might be able to make purge-old-deployments use some of the same cache-cleaning library code with a result of it being more readable; did you consider doing that?


Yeah I had considered doing so but hadn't found any "large" piece of code that can be de-duplicated. I had a bit confusion about the directory purging code on whether should it be put into library or not but after reading your comment and I went ahead and extracted a function for it. I think rest of the code is unique to `purge-old-deployments` script, is there anything else you feel can be moved to library.

HarshitOnGitHub commented on scripts/lib/zulip_tools.py:
```diff
@@ -264,3 +252,18 @@ def generate_sha1sum_emoji(zulip_path):
+    if dry_run:
+        print("Performing a dry run...")
+    else:
+        print("Cleaning unused %ss..." % (dir_type,))
```
I have used an extra `s` in the string to print `emoji caches` instead of `emoji cache`. Feel free to remove/modify it if it doesn't suit your taste.

HarshitOnGitHub commented:
@timabbott This is ready for a final sign off! :)

timabbott commented:
Looks great, merged, thanks @HarshitOnGitHub!  I'm testing it out on chat.zulip.org momentarily.
